### PR TITLE
Add an attention bias subclass for a lower right causal masking

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
This PR introduces a new Tensor subclass that is designed to be used with torch.nn.functional.scaled_dot_product_attention. Currently we have a boolean `is_causal` flag that allows users to do do causal masking without the need to actually create the "realized" attention bias and pass into sdpa. We originally added this flag since there is native support in both fused kernels we support. This provides a big performance gain ( the kernels only need to iterate over ~0.5x the sequence, and for very large sequence lengths this can provide vary large memory improvements.

The flag was introduced when the early on in the kernel development and at the time it was implicitly meant to "upper_left" causal attention. This distinction only matters when the attention_bias is not square. For a more detailed break down see: https://github.com/pytorch/pytorch/issues/108108. The kernels default behavior has since changed, largely due to the rise of autogressive text generation. And unfortunately this would lead to a BC break. In the long term it may actually be beneficial to change the default meaning of `is_causal` to represent lower_right causal masking.

The larger theme though is laid here: https://github.com/pytorch/pytorch/issues/110681. The thesis being that there is alot of innovation in SDPA revolving around the attention_bias being used. This is the first in hopefully a few more attention_biases that we would like to add. The next interesting one would be `sliding_window` which is used by the popular mistral model family. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114823

Results from benchmarking, I improved the meff_attention perf hence the slightly decreased max perf.
```Shell
+---------+--------------------+------------+-----------+-----------+-----------+-----------+----------------+----------+
|  Type   |      Speedup       | batch_size | num_heads | q_seq_len | k_seq_len | embed_dim |     dtype      | head_dim |
+---------+--------------------+------------+-----------+-----------+-----------+-----------+----------------+----------+
| Average | 1.2388050062214226 |            |           |           |           |           |                |          |
|   Max   | 1.831672915579016  |    128     |    32     |   1024    |   2048    |   2048    | torch.bfloat16 |    64    |
|   Min   | 0.9430534166730135 |     1      |    16     |    256    |    416    |   2048    | torch.bfloat16 |   128    |
+---------+--------------------+------------+-----------+-----------+-----------+-----------+----------------+----------+
```

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]

File: torch/nn/utils/attention.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        """ "Defines the behavior of torch.nn.functional.scaled_dot_product_attention when the attn_bias is an AttnBias"""
+        if kwargs is None:
+            kwargs = {}
+        if func != torch.nn.functional.scaled_dot_product_attention:
+            raise NotImplementedError(
+                "CausalBias only supports scaled_dot_product_attention"
+            )
+        return cls.dispatch(*args, **kwargs)
+
+    def __repr__(self) -> str:
```
Comment (cpuhrsch) [inline torch/nn/utils/attention.py:255]: I think this should just return f.materialize().__repr__()

File: torch/nn/utils/attention.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        self.variant = variant
+        self.seq_len_q = seq_len_q
+        self.seq_len_kv = seq_len_kv
+        if seq_len_q > seq_len_kv and variant == CausalVariant.LOWER_RIGHT:
+            warn(
+                "Lower right causal bias will produce NaNs in the output when seq_len_q > seq_len_kv!"
+            )
+
+    @staticmethod
+    def upper_left(seq_len_q: int, seqlen_kv: int) -> "CausalBias":
```
Comment (cpuhrsch) [inline torch/nn/utils/attention.py:145]: Is there another example of a class method factory function in PyTorch? If not I'd suggest we keep it in a namespace.

File: torch/nn/utils/attention.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        """Lower right causal bias"""
+        diagonal_offset = self.seq_len_kv - self.seq_len_q
+        return torch.tril(
+            torch.ones(
+                self.seq_len_q, self.seq_len_kv, device=device, dtype=torch.bool
+            ),
+            diagonal=diagonal_offset,
+        )
+
+    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
```
Comment (cpuhrsch) [inline torch/nn/utils/attention.py:188]: Why not to_dense?

File: torch/nn/utils/attention.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+        """
+        if device is None:
+            device = torch.device("cpu")
+        if self.variant == CausalVariant.UPPER_LEFT:
+            return self._upper_left(device)
+        elif self.variant == CausalVariant.LOWER_RIGHT:
+            return self._lower_right(device)
+
+    @staticmethod
+    def dispatch(
```
Comment (cpuhrsch) [inline torch/nn/utils/attention.py:209]: Does this need to be public for now?

File: torch/nn/functional.py [hunk 5]
```diff
@@ -5035,6 +5035,18 @@ def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.
     >>> with torch.backends.cuda.sdp_kernel(enable_math=False):
     >>>     F.scaled_dot_product_attention(query,key,value)
 
+    Using Attention Biases:
```
Comment (cpuhrsch) [inline torch/nn/functional.py:5038]: You could say that explicitly one of these options is equivalent to setting `is_causal=True`.

Comment (cpuhrsch) [reply torch/nn/functional.py:5038]: I'd also consider adding a testing that asserts that

File: torch/nn/utils/attention.py (hunk 2 above)
Comment (drisspg) [reply torch/nn/utils/attention.py:145]: I think that the name spacing is a actually more clear here:

```Python
from nn.utils.attention import CausalBias

bias = CausalBias.lower_right(...)
```
vs
```Python
from nn.utils.attention lower_right_causal_bias

bias = lower_right_causal_bias(...)
```
In either case both bias is going to be of the same type.  Let say a user wants to add typehinting then they will also need to import CausalBias

I would argue that this is just a more confined namespace. And going forward I think that we would suffer the same challenges w/ BC if its a function or not since the return type will need to change in either case no?

File: torch/nn/utils/attention.py (hunk 4 above)
Comment (drisspg) [reply torch/nn/utils/attention.py:209]: yeah good call this can be private

File: torch/nn/utils/attention.py (hunk 3 above)
Comment (drisspg) [reply torch/nn/utils/attention.py:188]: what if we punt on this for now and make this private as well?

File: torch/nn/attention/_utils.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
+"""Defines utilities for interacting with scaled_dot_product_attention"""
+import math
+from typing import List, Optional
+
+import torch
+
+__all__: List[str] = []
```
Comment (drisspg) [inline torch/nn/attention/_utils.py:7]: needed to appease mypy

File: torch/nn/functional.py [hunk 7]
```diff
@@ -5035,6 +5035,20 @@ def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.
...
+
+.. code-block:: python
+
+    from torch.nn.utils.attention import CausalBias
+
+    bsz, num_heads, seqlen_q, seqlen_kv, head_dim = 32, 8, 4, 12, 8
+    q = torch.randn(bsz, num_heads, seqlen_q, head_dim, device="cuda", dtype=torch.float16)
+    k = torch.randn(bsz, num_heads, seqlen_kv, head_dim, device="cuda", dtype=torch.float16)
+    v = torch.randn(bsz, num_heads, seqlen_kv, head_dim, device="cuda", dtype=torch.float16)
+    bias = CausalBias.lower_right(seqlen_q, seqlen_kv)
```
Comment (cpuhrsch) [inline torch/nn/functional.py:5048]: needs to be updated (just so you don't forget)

File: torch/nn/attention/bias.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+
+        [[1, 1, 0, 0],
+         [1, 1, 1, 0],
+         [1, 1, 1, 1]]
+
+    Args:
+        size: The size of the bias matrix.
+
+    Returns:
+        CausalBias: The upper-left triangular causal bias.
```
Comment (cpuhrsch) [inline torch/nn/attention/bias.py:306]: Maybe use the term `variant`?
